larray/tests/test_constrained_session.py

```diff
@@ -0,0 +1,703 @@
+import pytest
+import pickle
+
+import numpy as np
+
+from larray import (ConstrainedSession, ConstrainedArray, Axis, AxisCollection, Group, Array,
+                    ndtest, full, full_like, zeros_like, ones, ones_like, isnan)
+from larray.tests.common import (inputpath, tmp_path, assert_array_nan_equal, meta,
+                                 needs_pytables, needs_xlrd, needs_xlwings)
+from larray.tests.test_session import (a, a2, a3, anonymous, a01, ano01, b, b2, b024, c, d, e, f, g, h,
+                                       assertObjListEqual, session, test_getitem, test_getattr, test_add,
+                                       test_element_equals, test_eq, test_ne)
+from larray.core.constrained import NOT_LOADED
+
+
+class TestConstrainedSession(ConstrainedSession):
+    b = b
+    b024 = b024
+    a: Axis
+    a2: Axis
+    anonymous = anonymous
+    a01: Group
+    ano01 = ano01
+    c: str = c
+    d = d
+    e: Array
+    g: Array
+    f: ConstrainedArray((Axis(3), Axis(2)))
+    h: ConstrainedArray((a3, b2), dtype=int)
+
+
+@pytest.fixture()
+def constrainedsession():
+    return TestConstrainedSession(a=a, a2=a2, a01=a01, e=e, g=g, f=f, h=h)
+
+
+def test_create_constrainedsession_instance(meta):
+    # As of v1.0 of pydantic all fields with annotations (whether annotation-only or with a default value)
+    # will precede all fields without an annotation. Within their respective groups, fields remain in the
+    # order they were defined.
+    # See https://pydantic-docs.helpmanual.io/usage/models/#field-ordering
+    declared_variable_keys = ['a', 'a2', 'a01', 'c', 'e', 'g', 'f', 'h', 'b', 'b024', 'anonymous', 'ano01', 'd']
+
+    # setting variables without default values
+    cs = TestConstrainedSession(a, a01, a2=a2, e=e, f=f, g=g, h=h)
+    assert list(cs.keys()) == declared_variable_keys
+    assert cs.b.equals(b)
+    assert cs.b024.equals(b024)
+    assert cs.a.equals(a)
+    assert cs.a2.equals(a2)
+    assert cs.anonymous.equals(anonymous)
+    assert cs.a01.equals(a01)
+    assert cs.ano01.equals(ano01)
+    assert cs.c == c
+    assert cs.d == d
+    assert cs.e.equals(e)
+    assert cs.g.equals(g)
+    assert cs.f.equals(f)
+    assert cs.h.equals(h)
+
+    # metadata
+    cs = TestConstrainedSession(a, a01, a2=a2, e=e, f=f, g=g, h=h, meta=meta)
+    assert cs.meta == meta
+
+    # override default value
+    b_alt = Axis('b=b0..b4')
+    cs = TestConstrainedSession(a, a01, b=b_alt, a2=a2, e=e, f=f, g=g, h=h)
+    assert cs.b is b_alt
+
+    # test for "NOT_LOADED" variables
+    not_loaded_variables = ['a', 'a2', 'a01', 'e', 'g', 'f', 'h']
+    with pytest.warns(UserWarning) as caught_warnings:
+        cs = TestConstrainedSession()
+    assert len(caught_warnings) == len(not_loaded_variables)
+    for i, var_name in enumerate(not_loaded_variables):
+        assert caught_warnings[i].message.args[0] == f"No value passed for the declared variable '{var_name}'"
```

Comment:
I **think** a warning will be annoying. I think we should either allow this silently or raise an Exception. I don't know which.

Reply:
If I remember well the problem with raising an Exception was that any code trying to build a `ConstrainedSession` by loading several files will fail (assuming the variables of such session are stored in separated input files).
   
If we choose to allow this silently, my biggest fear is that users may face obscure error messages when they will try to make any operation on a not loaded variable.

We might first allow this silently and then wait for users to complain.

Reply:
What about replacing
```python
NOT_LOADED = object()
```
by 
```python
class NOT_LOADED(object):
    pass
```
?

This can make error messages when using unloaded variables more explicit (at least for both of us if we need to help them to debug).

Reply:
I am fine with any solution as long as we give ourselve the opportinity to revisit this choice later. Makes me think that we might want to label the new feature "experimental" for one or two releases.